lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,192 @@ def as_dict_with_promql_labels(self):
         and topology identifier.
 
         Args:
-            path: path to rule file.
+            root_path: path to the root rules dir.
+            file_path: path to rule file.
+            group_name: original group name to keep as part of the new augmented group name
+
+        Returns:
+            New group name, augmented by juju topology and relative path.
         """
-        relpath = os.path.relpath(os.path.dirname(path), str(dir_path))
+        rel_path = os.path.relpath(os.path.dirname(file_path), root_path)
+        rel_path = "" if rel_path == "." else rel_path.replace(os.path.sep, "_")
 
         # Generate group name:
-        #  - prefix, from the relative path of the rule file;
         #  - name, from juju topology
-        prefix = "" if relpath == "." else relpath.replace(os.path.sep, "_") + "_"
-        return "{}{}_alerts".format(prefix, topology.identifier)
+        #  - suffix, from the relative path of the rule file;
+        group_name_parts = [self.topology.identifier, rel_path, group_name, "alerts"]
+        return "_".join(filter(None, group_name_parts))
+
+    def _from_dir(self, dir_path: Path, recursive: bool) -> List[dict]:
+        """Read all rule files in a directory.
+
+        All rules from files for the same directory are loaded into a single
+        group. The generated name of this group includes juju topology.
+        By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+
+        Args:
+            dir_path: directory containing *.rule files (alert rules without groups).
+            recursive: flag indicating whether to scan for rule files recursively.
+
+        Returns:
+            a list of dictionaries representing prometheus alert rule groups, each dictionary
+            representing an alert group (structure determined by `yaml.safe_load`).
+        """
+        alert_groups = []  # type: List[dict]
 
-    paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")  # type: ignore
-    for path in filter(Path.is_file, paths):
-        if rule := load_alert_rule_from_file(path, topology):
-            logger.debug("Reading alert rule from %s", path)
-            alerts[_group_name(path)].append(rule)
+        # Gather all alerts into a list of groups
+        paths = dir_path.glob("**/*.rule" if recursive else "*.rule")
+        for file_path in filter(Path.is_file, paths):
+            alert_groups_from_file = self._from_file(dir_path, file_path)
+            if alert_groups_from_file:
+                logger.debug("Reading alert rule from %s", file_path)
+                alert_groups.extend(alert_groups_from_file)
 
-    # Gather all alerts into a list of groups since Prometheus
-    # requires alerts be part of some group
-    groups = [{"name": k, "rules": v} for k, v in alerts.items()]
-    return groups
+        return alert_groups if alert_groups else []
```

Comment:
```suggestion
        return alert_groups
```
since if expression is redundant given `alert_groups` is correctly initialized.